### PR TITLE
Add some options to GenerateRuntimeGraph for testing

### PR DESCRIPTION
**Pull request by ericstj**
This adds support for emitting a DGML and a compatibility mapping from a runtime graph.

The DGML is handy for visualizing the runtime graph and the compatibility mapping
helps identify what the effects of a RID graph change will be on package asset resolution.

/cc @weshaggard @eerhardt

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs [hunk 1]
```diff
@@ -75,21 +75,38 @@ public ITaskItem SourceRuntimeJson
...
         public ITaskItem RuntimeJson
         {
             get;
             set;
         }
 
+        /// <summary>
+        /// When defined, specifies the file to write compatibility precedence for each RID in the graph.
+        /// </summary>
+        public ITaskItem CompatibilityMap
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: Any reason why these are TaskItems instead of just `string` properties?

Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: Nope, I can make them strings.

**Comment by ericstj** [comment]
Using this I've actually found some compat issues with the way I'm generating the RID heirarchy for windows.  I'm going to roll in a fix for that to this PR.  Stay tuned.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs [hunk 2]
```diff
@@ -33,88 +35,81 @@ public class GenerateRuntimeGraph : PackagingTask
...
         ///     exponential growth of RIDs).
+        ///
+        ///   *Use at your own risk*
         ///   OmitRIDs: A list of strings delimited by semi-colons that represent RIDs calculated from this RuntimeGroup that should
-        ///     be omitted from the RuntimeGraph.  This is useful in cases where overlapping RuntimeGroups are used and one is 
-        ///     designated to be "best" for some set of calculated RIDs, the others may omit the overlapping RIDs.
+        ///     be omitted from the RuntimeGraph.  These RIDs will not be referenced nor defined.
+        ///   OmitRIDDefinitions: A list of strings delimited by semi-colons that represent RIDs calculated from this RuntimeGroup 
+        ///     that should be omitted from the RuntimeGraph.  These RIDs will not be defined by this runtimegroup, but will be 
+        ///     refereenced: usefull in case some other RuntimeGroup (or runtime.json template) defines them.
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: `refereenced`

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs [hunk 3]
```diff
@@ -33,88 +35,81 @@ public class GenerateRuntimeGraph : PackagingTask
         ///   AdditionalQualifiers: A list of strings delimited by semi-colons that represent the additional qualifiers for this RID.
         ///     Additional qualifers do not stack, each only applies to the qualifier-less RIDs (so as not to cause combinatorial 
         ///     exponential growth of RIDs).
+        ///
+        ///   *Use at your own risk*
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: I'm not sure I understand this warning.  Can it be explained further?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs [hunk 4]
```diff
@@ -194,6 +190,8 @@ public RuntimeGroup(ITaskItem item)
                 Architectures = item.GetStrings(nameof(Architectures));
                 AdditionalQualifiers = item.GetStrings(nameof(AdditionalQualifiers));
                 OmitRIDs = new HashSet<string>(item.GetStrings(nameof(OmitRIDs)));
+                OmitRIDs = new HashSet<string>(item.GetStrings(nameof(OmitRIDDefinitions)));
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: I think this is supposed to write to `OmitRIDDefinitions` and `OmitRIDReferences`, not overwrite OmitRIDs each time.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs [hunk 5]
```diff
@@ -205,137 +203,336 @@ public RuntimeGroup(ITaskItem item)
...
-                            imports.Add(runtimeDescription.RuntimeIdentifier);
+                            lastVersion = version;
                         }
+                    }
+                }
+            }
+
+            private bool IsNullOrRoot(string rid)
+            {
+                return rid == null || rid == rootRID;
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: I'm a bit confused at the "rootRID".  In the code above, it is "any".  But in the [runtime.json file](https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/runtime.json), it seems that "any" imports "base".

I would think "base" is the rootRID.

Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: `any` is the true root.  `base` was introduced as a private RID that only we would use back when we tried to use runtime.json to lift up our package versions to a minimum version https://github.com/dotnet/buildtools/blob/96ee49b1f47f1a61dc040b2be2d976bbaa7903c4/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs#L195-L207.  This technique never really worked perfectly so we stopped using it, then NuGet broke it (intentionally) by allowing downgrade of package versions referenced by runtime.json if the version referenced was higher in the graph than the package which brought in the runtime package.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs (hunk 3 above)
Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: Omitting RIDs from the graph can breaks the compatibility precedence we ensure when generating RIDs/imports from RuntimeGroups.  This makes it easy to create a RID fallback chain that doesn't match the rest of the RIDs and therefore is hard for developers/package authors to reason about.  Only use these options for cases where you know what you are doing and have carefully reviewed the resulting RID fallbacks.

Comment (eerhardt) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: Sounds like a perfect comment to put into this code 😉

Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: I'll add a comment to the source.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs (hunk 4 above)
Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs]: Doh, good catch

**Comment by ericstj** [comment]
> If in the future we need more precedences we'll have to refactor this, one thing to watch
out for in particular is that if version doesn't have the highest precedence we'll have to
flatten the version list to imports for a single RID for compatible versioned groups.  That
will not scale well which is another reason I've avoided exposing such functionality.

I was worried that we'd get this wrong if we tried to do it later.  I went ahead and brought back QualifierPolicies and implemented qualifier-first-precedence, since it seems closer to what corert was trying to do before.  I inverted the default since using qualifier-first-precedence is not necessarily something we want to do everywhere due to how big it grows the runtime graph.

**Comment by ericstj** [comment]
Ok, that's a failed experiment.  It's not as simple as just saying qualifiers have precedence.  Since the RIDs with qualifiers must also import the RIDs without qualifiers (in order to associate at the right os/version/arch) and runtime-compatibility is a breadth first traversal, it means that qualified RIDs must contain their *entire* flattened qualifier specific compatibility mappings (all the way up the graph) as ordered imports.  This is not scalable in a static representation so I'm cutting it.

**Comment by ericstj** [comment]
You'll probably want to review the diffs with this change: https://github.com/dotnet/corefx/pull/24260

**Pull request merged**